### DISCUSSION TITLE
Is it possible to get the search result score?

### DISCUSSION THREAD

**Discussion by FourierTransformer** [meilisearch/product#602] [Feedback & Feature Proposal]
I'm curious if I can get the internal ranking/score when performing a search.

**Comment by Kerollmops** [comment]
Hey @FourierTransformer,

Yeah, this is planed and is the 6th step of https://github.com/meilisearch/MeiliSearch/issues/692.

**Reply by FourierTransformer** [reply]
That sounds great, thanks! Meilisearch helped make adding search to my site really easy, and the search results are really good! Thanks for all the work the team is putting in!

**Comment by raine** [comment]
Would this be part of a solution to being able to filter out only remotely accurate search results from otherwise accurate?

For example, I have a case where 6 of the results contain the search keywords in the title field, but there's also 3 other results that have them somewhere in the fields.

I would imagine you could use the disparity in score to filter out the other three.

**Reply by tpayet** [reply]
Hi @raine, you might want to have a look into the ranking system. You should be able to filter out the three results that got the query keywords not in title by removing those fields in the searchable attributes: https://docs.meilisearch.com/guides/main_concepts/relevancy.html#changing-the-order-of-the-attributes

**Comment by JonathonWebster** [comment]
Hi @Kerollmops,

Would this return a numerical value for each ranking rule? I'm hoping to use the numerical values to create a custom rank based on the existing rules.  This would allow me to have a less linear ranking rule priority (if i say typos is the number 1 priority, and number of words matching 2nd priority, something with only 1 word missing might still be better that something with 5 typos).  

I haven't started playing with this directly yet, but after reviewing the documentation, it seems like I would need something like this in order to optimize my ecommerce product feed.  A more likely example would be needing to weigh estimated revenue and overall relevance - in which case I would still need to design a consolidated 'relevance' score.

Lastly, in the first example I mentioned, do the ranking rules truly work like this, where if I choose typos ranked as #1, it will give me all results with fewest typos, regardless of how well the other rules perform?

Thanks for your help!

**Comment by MihaiNueleanu** [comment]
Hi ! It's been a while since the question was originally asked. Is this feature still planned for development?

**Reply by curquiza** [reply]
Hello, the feature suggestions are now handled on the product repo :) here is the dedicated discussion about it, I've just opened it since it looks like it has not been created yet https://github.com/meilisearch/product/discussions/379
It has not been prioritized and is not planned since not a lot of users requested it so far compared to other features.

If you have time, you can share your need and your usecase. It would be definitely helpful to know what you need and to maybe prioritize it 🙂

**Comment by ar-siddiqui** [comment]
I would like this feature to get search results of different indexes and then merge and sort them using relevancy score.

**Reply by curquiza** [reply]
Hello @ar-siddiqui thanks for your input, we plan to develop this feature, we cannot give any ETA at the moment, but it's planned.

@gmourier FYI

**Reply by Nyholm** [reply]
Is there anything I can to to help?

**Reply by curquiza** [reply]
Hello @Nyholm 
thanks for suggesting to paritcipate 😄 
I've asked the team, if you say you are still interested in doing the feature, we'll come back to you. We indeed don't know exactly what and how we want to implement this

By curiosity, are you asking because you really need this feature, and if yes, how would you like to use this feature?
If no, I mean, if you are asking this to contribute to an opensource product, let me know about your favorite languages, I will redirect you to other issues we would need to solve!

**Reply by Nyholm** [reply]
Hey. 
Im asking because I really want this feature.  =)
I am using multiple metrics when searching to add a personal "recommendation score". Say that you are searching in a database for candidates to a new job. The user types in "dog walker cleaning homework" because they want to find a resume for a person who can help walking their dog etc. I want to combine this Meilisearch result with scores like 
- "Have you been working with this person before"
- "Is the wage close to your salary range"
- "Boost if they have made many jobs before"
- "Boost if they have a good rating". 

I do know that I can add custom rankings. But that is (A) very complex, and (B) it does not allow my user to sort on search relevance without reloading the page. 

I also would like to combine the result with another (more implementation specific) search engine. 

---------

> If no, I mean, if you are asking this to contribute to an opensource product, let me know about your favorite languages, I will redirect you to other issues we would need to solve!

I am a fairly good PHP developer. I see that the [meilisearch-php](https://github.com/meilisearch/meilisearch-php) client is well written and [its only dependencies are packages that I maintain](https://github.com/meilisearch/meilisearch-php/blob/v0.24.2/composer.json#L17-L19). Some other [test dependencies](https://github.com/meilisearch/meilisearch-php/blob/v0.24.2/composer.json#L38-L39) are also maintained by me. I also appreciate that your readme recommend `nyholm/psr7`

**Reply by gmourier** [reply]
Hey @Nyholm 👋

Thanks for the answer.

> I also would like to combine the result with another (more implementation specific) search engine.

Do you have a name to share? I'm super curious. Will this search engine do the boosting given the users' profile data? 

We have already discussed this subject with the core team; today, it is very difficult, if not impossible, to return a normalized score between 0 and 1.

What we could eventually do would be [something similar to what Algolia](https://www.algolia.com/doc/api-reference/api-methods/search/#method-response-_rankinginfo) returns. WDYT?

**Reply by utkarshiam** [reply]
> Hey @Nyholm 👋
> 
> Thanks for the answer.
> 
> > I also would like to combine the result with another (more implementation specific) search engine.
> 
> Do you have a name to share? I'm super curious. Will this search engine do the boosting given the users' profile data?
> 
> We have already discussed this subject with the core team; today, it is very difficult, if not impossible, to return a normalized score between 0 and 1.
> 
> What we could eventually do would be [something similar to what Algolia](https://www.algolia.com/doc/api-reference/api-methods/search/#method-response-_rankinginfo) returns. WDYT?

That would be awesome! 
Anything which gives me score to so an analysis over the search results from Meili would do the job.

**Reply by Nyholm** [reply]
Yeah. That would be great. 

> Do you have a name to share?

We use a custom TF-IDF and a plain database search.

**Reply by Snarik** [reply]
Im just going to chime in and say I would also very much like a normalized score with the hits.

**Reply by Kerollmops** [reply]
Hey everyone 👋,

We are looking into this feature as we are working on returning the ranking rules scores in a future version of the engine.
Would the ranking rule score be enough/adapted for your use cases?

The ranking rule score corresponds to the score each ranking rule will return in the `rankingRules` settings order.
I understand that it can be hard to order an array of scores especially if you don't know if a certain element must be ordered in descending or ascending order, we will take that into account to expose a better API.
Here are some settings examples and the corresponding score that could be returned for each document:

```json
{
  "rankingRules": [
    "words",
    "typo",
    "name:desc",
  ]
}
```

```json5
[
  {
    "document": {
      "title": "I am the best document",
      "name": "kero"
    },
    "score": [
      3, // matches 3 words of the query
      0, // no typos
      "kero" // we can't compute a normilized score for a string so we put it here.
    ]
  },
  {
    "document": {
      "title": "I am the second document",
      "name": "pierro"
    },
    "score": [
      3, // matches 3 words of the query
      1, // one typos, that's what moved this document down
      "pierro"
    ]
  }
]
```

About the score ordering we could return an array explaining the order of each ranking rule and their names:
```json
[
  { "name": "words", "ordering": "desc" },
  { "name": "typos", "ordering": "desc" },
  { "name": "name:desc", "ordering": "desc" }
]
```

**Reply by inspire22** [reply]
That would be very helpful, thanks :) A simplified "rank" score (just the other ones summed?) would match what mysql's fulltext search provides, but we could always calculate that manually from the full list of sub-scores.  I'm using meilisearch to detect plagarism using fuzzy searches, then for high-scoring results doing a Levenshtein distance or ML driven more complicated final ranking from that subset (if needed, the score might actually be enough).  Meilisearch seems awesome!  Are there beta branches followable to try stuff like this out early?

**Reply by gmourier** [reply]
Hello @inspire22 👋

> A simplified "rank" score (just the other ones summed?) would match what mysql's fulltext search provides, but we could always calculate that manually from the full list of sub-scores.

Do you mean having a score between 0 and 1 ?

> Are there beta branches followable to try stuff like this out early?

Not yet! But we will work on it in Q2/Q3 😇 We will keep everyone updated whenever we move on that

**Comment by inspire22** [comment]
Mysql's fulltext score seems to be an arbitrary number, I'm not sure if there's a maximum.  For me it was usually between 0 and 70 or so.  Whatever works.  It can be nice to know when the top result (or top 3) is 50x stronger of a match than the next match, for example.

**Comment by macraig** [comment]
Hello everyone! 👋

Quick update on our side, we've started working on a solution to return ranking score details on search.

You can follow the updates in https://github.com/meilisearch/product/discussions/594

**Comment by macraig** [comment]
**Update:** Scoring details will be available as an **experimental feature** in Meilisearch `v1.3` (released July 31st)

You can find the details in https://github.com/meilisearch/product/discussions/674

**Reply by Snarik** [reply]
extremely exciting!

**Comment by macraig** [comment]
Hello everyone 👋 

We have just released the first RC (release candidate) of Meilisearch containing this new feature!
You can test it by using
-  the [release assets](https://github.com/meilisearch/meilisearch/releases)
- the [Meilisearch Docker image](https://hub.docker.com/repository/docker/getmeili/meilisearch/general)

```bash
docker run -it --rm -p 7700:7700 -v $(pwd)/meili_data:/meili_data getmeili/meilisearch:v1.3.0-rc.0
```

You are welcome to leave your feedback in this discussion.

If you encounter any bugs, please report them [here](https://github.com/meilisearch/meilisearch/issues/new?assignees=&labels=&template=bug_report.md&title=).
Thanks in advance for your help and your involvement in Meilisearch ❤️ 

🎉 The official and stable release containing this change will be available on July 31st, 2023

⚠️ RC (release candidates) are not recommended for production

**Comment by macraig** [comment]
Hey folks 👋 

[v1.3](https://github.com/meilisearch/meilisearch/releases/tag/v1.3.0) has been released! 🦁  You can now get global and detailed ranking scores for your documents ✨

**Note:** Document ranking score details are considered Experimental, you can find the usage instructions in https://github.com/meilisearch/product/discussions/674

📚 https://www.meilisearch.com/docs/learn/core_concepts/relevancy#ranking-score
📚 https://www.meilisearch.com/docs/learn/experimental/ranking_rule_score_details
